File: inventario_visual/DAO/FacturaDAO.vb
Imports MySql.Data.MySqlClient
Imports System.Data

Namespace inventario_visual
    Public Class FacturaDAO
        Public Function ListarTodas() As List(Of Factura)
            Dim lista As New List(Of Factura)
            Using conn = ConexionDB.GetConnection()
                conn.Open()
                Dim sql As String = "SELECT f.*, c.nombre AS cliente_nombre FROM factura f LEFT JOIN cliente c ON f.cliente_id=c.id ORDER BY f.id DESC"
                Using cmd As New MySqlCommand(sql, conn), reader = cmd.ExecuteReader()
                    While reader.Read()
                        lista.Add(Map(reader))
                    End While
                End Using
            End Using
            Return lista
        End Function

        Public Function Buscar(texto As String) As List(Of Factura)
            Dim lista As New List(Of Factura)
            Using conn = ConexionDB.GetConnection()
                conn.Open()
                Dim enc As String = AESUtil.Encriptar(texto)
                Dim sql As String = "SELECT f.*, c.nombre AS cliente_nombre FROM factura f LEFT JOIN cliente c ON f.cliente_id=c.id WHERE f.numero_factura LIKE @t OR f.numero_factura = @enc ORDER BY f.id DESC"
                Using cmd As New MySqlCommand(sql, conn)
                    cmd.Parameters.AddWithValue("@t", "%" & texto & "%")
                    cmd.Parameters.AddWithValue("@enc", enc)
                    Using reader = cmd.ExecuteReader()
                        While reader.Read()
                            lista.Add(Map(reader))
                        End While
                    End Using
                End Using
            End Using
            Return lista
        End Function

        Public Function GenerarNumero() As String
            Dim ultimoNumero As String = ""
            Using conn = ConexionDB.GetConnection()
                conn.Open()
                Using cmd As New MySqlCommand("SELECT numero_factura FROM factura ORDER BY id DESC LIMIT 1", conn)
                    Dim result = cmd.ExecuteScalar()
                    If result IsNot Nothing AndAlso Not IsDBNull(result) Then
                        ultimoNumero = AESUtil.Desencriptar(result.ToString())
                    End If
                End Using
            End Using
            Dim nextNum As Integer = 1
            If Not String.IsNullOrEmpty(ultimoNumero) AndAlso ultimoNumero.StartsWith("FAC-") Then
                Dim parts() As String = ultimoNumero.Split("-"c)
                If parts.Length >= 2 AndAlso Integer.TryParse(parts(parts.Length - 1), nextNum) Then
                    nextNum += 1
                End If
            End If
            Return String.Format("FAC-{0:D3}", nextNum)
        End Function

        Public Sub Insertar(f As Factura)
            Using conn = ConexionDB.GetConnection()
                conn.Open()
                Dim sql As String = "INSERT INTO factura (movimiento_id,cliente_id,numero_factura,metodo_pago,estado,subtotal,impuestos,total) VALUES (@mov,@cli,@num,@metodo,@estado,@subtotal,@impuestos,@total)"
                Using cmd As New MySqlCommand(sql, conn)
                    cmd.Parameters.AddWithValue("@mov", If(f.MovimientoId.HasValue, CObj(f.MovimientoId.Value), DBNull.Value))
                    cmd.Parameters.AddWithValue("@cli", If(f.ClienteId.HasValue, CObj(f.ClienteId.Value), DBNull.Value))
                    cmd.Parameters.AddWithValue("@num", AESUtil.Encriptar(f.NumeroFactura))
                    cmd.Parameters.AddWithValue("@metodo", AESUtil.Encriptar(f.MetodoPago))
                    cmd.Parameters.AddWithValue("@estado", AESUtil.Encriptar(f.Estado))
                    cmd.Parameters.AddWithValue("@subtotal", f.Subtotal)
                    cmd.Parameters.AddWithValue("@impuestos", f.Impuestos)
                    cmd.Parameters.AddWithValue("@total", f.Total)
                    cmd.ExecuteNonQuery()
                End Using
            End Using
        End Sub

        Public Sub ActualizarEstado(id As Long, estado As String)
            Using conn = ConexionDB.GetConnection()
                conn.Open()
                Using cmd As New MySqlCommand("UPDATE factura SET estado=@estado WHERE id=@id", conn)
                    cmd.Parameters.AddWithValue("@estado", AESUtil.Encriptar(estado))
                    cmd.Parameters.AddWithValue("@id", id)
                    cmd.ExecuteNonQuery()
                End Using
            End Using
        End Sub

        Public Sub Eliminar(id As Long)
            Using conn = ConexionDB.GetConnection()
                conn.Open()
                Using cmd As New MySqlCommand("DELETE FROM factura WHERE id=@id", conn)
                    cmd.Parameters.AddWithValue("@id", id)
                    cmd.ExecuteNonQuery()
                End Using
            End Using
        End Sub

        Private Shared Function Map(reader As MySqlDataReader) As Factura
            Return New Factura With {.Id = Convert.ToInt64(reader("id")), .MovimientoId = If(IsDBNull(reader("movimiento_id")), CType(Nothing, Long?), Convert.ToInt64(reader("movimiento_id"))), .ClienteId = If(IsDBNull(reader("cliente_id")), CType(Nothing, Long?), Convert.ToInt64(reader("cliente_id"))), .NumeroFactura = AESUtil.Desencriptar(reader("numero_factura").ToString()), .FechaEmision = Convert.ToDateTime(reader("fecha_emision")), .MetodoPago = AESUtil.Desencriptar(If(IsDBNull(reader("metodo_pago")), "", reader("metodo_pago").ToString())), .Estado = AESUtil.Desencriptar(If(IsDBNull(reader("estado")), "", reader("estado").ToString())), .Subtotal = Convert.ToDouble(reader("subtotal")), .Impuestos = Convert.ToDouble(reader("impuestos")), .Total = Convert.ToDouble(reader("total")), .ClienteNombre = AESUtil.Desencriptar(If(IsDBNull(reader("cliente_nombre")), "", reader("cliente_nombre").ToString()))}
        End Function
    End Class
End Namespace

File: inventario_python/dao/factura_dao.py
from db.conexion import conectar
from models.factura import Factura
from utils.aes_util import encriptar, desencriptar


class FacturaDAO:

    def _mapear(self, row):
        f = Factura()
        f.id = row[0]
        f.movimiento_id = row[1]
        f.cliente_id = row[2]
        f.numero_factura = desencriptar(row[3]) if row[3] else ""
        f.fecha_emision = row[4]
        f.metodo_pago = desencriptar(row[5]) if row[5] else ""
        f.estado = desencriptar(row[6]) if row[6] else ""
        f.subtotal = row[7] or 0.0
        f.impuestos = row[8] or 0.0
        f.total = row[9] or 0.0
        if len(row) > 10 and row[10]:
            f.cliente_nombre = desencriptar(row[10])
        return f

    def listar_todas(self):
        conn = conectar()
        if not conn:
            return []
        try:
            cur = conn.cursor()
            cur.execute("SELECT * FROM factura ORDER BY id DESC")
            return [self._mapear(r) for r in cur.fetchall()]
        finally:
            conn.close()

    def listar_todas_con_cliente(self):
        conn = conectar()
        if not conn:
            return []
        try:
            cur = conn.cursor()
            cur.execute(
                "SELECT f.*, c.nombre as cliente_nombre FROM factura f "
                "LEFT JOIN cliente c ON f.cliente_id = c.id ORDER BY f.id DESC"
            )
            return [self._mapear(r) for r in cur.fetchall()]
        finally:
            conn.close()

    def buscar(self, texto):
        conn = conectar()
        if not conn:
            return []
        try:
            cur = conn.cursor()
            enc = encriptar(texto)
            cur.execute("SELECT * FROM factura WHERE numero_factura LIKE %s OR numero_factura = %s ORDER BY id DESC", (f"%{texto}%", enc))
            return [self._mapear(r) for r in cur.fetchall()]
        finally:
            conn.close()

    def generar_numero(self):
        conn = conectar()
        if not conn:
            return "FAC-001"
        try:
            cur = conn.cursor()
            cur.execute("SELECT numero_factura FROM factura ORDER BY id DESC LIMIT 1")
            row = cur.fetchone()
            if row and row[0]:
                num = desencriptar(row[0])
                try:
                    n = int(num.split("-")[1]) + 1
                    return f"FAC-{n:03d}"
                except Exception:
                    return "FAC-001"
            return "FAC-001"
        finally:
            conn.close()

    def insertar(self, f):
        conn = conectar()
        if not conn:
            return False
        try:
            cur = conn.cursor()
            cur.execute(
                "INSERT INTO factura (movimiento_id, cliente_id, numero_factura, fecha_emision, metodo_pago, estado, subtotal, impuestos, total) "
                "VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s)",
                (f.movimiento_id, f.cliente_id, encriptar(f.numero_factura) if f.numero_factura else None,
                 f.fecha_emision, encriptar(f.metodo_pago) if f.metodo_pago else None,
                 encriptar(f.estado) if f.estado else None, f.subtotal, f.impuestos, f.total),
            )
            conn.commit()
            return True
        except Exception as e:
            conn.rollback()
            raise e
        finally:
            conn.close()

    def actualizar_estado(self, id, estado):
        conn = conectar()
        if not conn:
            return False
        try:
            cur = conn.cursor()
            cur.execute("UPDATE factura SET estado=%s WHERE id=%s", (encriptar(estado), id))
            conn.commit()
            return True
        except Exception as e:
            conn.rollback()
            raise e
        finally:
            conn.close()

    def eliminar(self, id):
        conn = conectar()
        if not conn:
            return False
        try:
            cur = conn.cursor()
            cur.execute("DELETE FROM factura WHERE id = %s", (id,))
            conn.commit()
            return True
        except Exception as e:
            conn.rollback()
            raise e
        finally:
            conn.close()
